tools/generator/autorest_ext/changelog_ext/changelog.go

```diff
@@ -0,0 +1,442 @@
+// A good part of rewriting everything is that we could keep these functions consistent
+
+func GetPackagesReportFromContent(lhs RepoContent, targetRoot string) (*PkgsReport, error) {
+	rhs, err := GetRepoContent(targetRoot)
+	if err != nil {
+		return nil, err
+	}
+	r := GetPkgsReport(lhs, rhs)
+	return &r, nil
+}
+
+// RepoContent contains repo content, it's structured as "package path (relative to the root of sdk)":content
+type RepoContent map[string]exports.Content
+
+func GetRepoContent(sdkRoot string) (RepoContent, error) {
+	// we must list over the services directory, otherwise it would walk into the .git directory and panic out
+	pkgs, err := track1.List(sdk.ServicesPath(sdkRoot))
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := getExportsForPackages(pkgs, sdkRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// returns repoContent based on the provided slice of package directories
+func getExportsForPackages(pkgs []track1.Package, root string) (RepoContent, error) {
+	exps := RepoContent{}
+	for _, pkg := range pkgs {
+		relativePath, err := filepath.Rel(root, pkg.FullPath())
+		if err != nil {
+			return nil, err
+		}
+		relativePath = utils.NormalizePath(relativePath)
+		if _, ok := exps[relativePath]; ok {
+			return nil, fmt.Errorf("duplicate package: %s", pkg.Path())
+		}
+		exp, err := exports.Get(pkg.FullPath())
+		if err != nil {
+			return nil, err
+		}
+		exps[relativePath] = exp
+	}
+	return exps, nil
+}
+
+const (
+	sdkRoot            = "github.com/Azure/azure-sdk-for-go"
+	absoluteLinkPrefix = "https://github.com/Azure/azure-sdk-for-go/tree/master/"
+)
+
+// contains a collection of packages
+type PkgsList []string
+
+// contains a collection of package reports, it's structured as "package path":pkgReport
+type ModifiedPackages map[string]report.Package
+
+// represents a complete report of added, removed, and modified packages
+type PkgsReport struct {
+	// AddedPackages contains the relative package names that are added in this new version
+	AddedPackages PkgsList `json:"added,omitempty"`
+	// ModifiedPackages contains the modified packages map, using relative package names as keys
+	ModifiedPackages ModifiedPackages `json:"modified,omitempty"`
+	// RemovedPackages contains the relative package names that are removed in this new version
+	RemovedPackages    PkgsList `json:"removed,omitempty"`
+	modPkgHasAdditions bool
+	modPkgHasBreaking  bool
+}
+
+// returns true if the package report contains breaking changes
+func (r PkgsReport) hasBreakingChanges() bool {
+	return len(r.RemovedPackages) > 0 || r.modPkgHasBreaking
+}
+
+// returns true if the package report contains additive changes
+func (r PkgsReport) hasAdditiveChanges() bool {
+	return len(r.AddedPackages) > 0 || r.modPkgHasAdditions
+}
+
+// returns true if the report contains no data
+func (r PkgsReport) isEmpty() bool {
+	return len(r.AddedPackages) == 0 && len(r.ModifiedPackages) == 0 && len(r.RemovedPackages) == 0
+}
```

Comment:
We have a lot of same code in `tools/internal` exactly like this - could we please refactor the code here to use existing functions instead?

Reply:
merged with internal/report/packages.go